File: go/providers/temp_mail_io.go
package providers

import (
	"bytes"
	"encoding/json"
	"net/http"
	"time"
)

const tempMailIoBaseURL = "https://api.internal.temp-mail.io/api/v3"

// TempMailIo is the temp-mail.io provider (REST API, Bearer token).
type TempMailIo struct {
	client *HttpClient
	token  string
	email  string
}

// NewTempMailIo creates a new TempMailIo provider.
func NewTempMailIo(config map[string]string) *TempMailIo {
	proxies := ParseProxies(config["proxies"])
	randomUA := config["randomUa"] == "true"
	return &TempMailIo{client: NewHttpClient(proxies, randomUA, false)}
}

// authReq builds a request with the Bearer token and routes it through HttpClient.
func (t *TempMailIo) authReq(method, path string, body []byte) ([]byte, error) {
	req, _ := http.NewRequest(method, tempMailIoBaseURL+path, bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	if t.token != "" {
		req.Header.Set("Authorization", "Bearer "+t.token)
	}
	return t.client.RawDo(req, body)
}

func (t *TempMailIo) GenerateEmail() (string, error) {
	payload, _ := json.Marshal(map[string]int{"min_name_length": 6, "max_name_length": 12})
	body, err := t.authReq(http.MethodPost, "/email/new", payload)
	if err != nil {
		return "", &TempMailError{Provider: "temp-mail.io", Message: err.Error()}
	}
	var resp struct {
		Email string `json:"email"`
		Token string `json:"token"`
	}
	if e := json.Unmarshal(body, &resp); e != nil || resp.Email == "" {
		return "", &TempMailError{Provider: "temp-mail.io", Message: "no email in response"}
	}
	t.email = resp.Email
	t.token = resp.Token
	return t.email, nil
}

func (t *TempMailIo) GetInbox(email string) ([]Message, error) {
	items, err := t.fetchMessages()
	if err != nil {
		return nil, err
	}
	msgs := make([]Message, len(items))
	for i, it := range items {
		id := rawString(it.ID)
		if id == "" {
			id = rawString(it.UID)
		}
		date := it.CreatedAt
		if date == "" {
			date = it.Date
		}
		msgs[i] = Message{ID: id, Sender: senderFrom(it.From), Subject: it.Subject, Date: parseMailDate(date)}
	}
	return msgs, nil
}

func (t *TempMailIo) ReadMessage(messageID string) (*MessageDetail, error) {
	// No dedicated read endpoint; the messages list already carries full bodies.
	items, err := t.fetchMessages()
	if err != nil {
		return nil, err
	}
	for _, it := range items {
		id := rawString(it.ID)
		if id == "" {
			id = rawString(it.UID)
		}
		if id == messageID {
			date := it.CreatedAt
			if date == "" {
				date = it.Date
			}
			text := it.BodyText
			if text == "" {
				text = it.Text
			}
			html := it.BodyHTML
			if html == "" {
				html = it.HTML
			}
			detail := &MessageDetail{
				Message: Message{
					ID:      messageID,
					Sender:  senderFrom(it.From),
					Subject: it.Subject,
					Date:    parseMailDate(date),
				},
				BodyText: text,
				BodyHTML: html,
			}
			for _, a := range it.Attachments {
				detail.Attachments = append(detail.Attachments, map[string]any{
					"filename":     a.Filename,
					"content_type": a.ContentType,
					"size":         a.Size,
				})
			}
			return detail, nil
		}
	}
	return nil, &NotFoundError{Provider: "temp-mail.io", Resource: "message " + messageID}
}

func (t *TempMailIo) DeleteEmail(email string) (bool, error) {
	t.token = ""
	t.email = ""
	return true, nil
}

func (t *TempMailIo) WaitForEmail(email string, timeout time.Duration, interval time.Duration) (*Message, error) {
	return waitForEmail(t, email, timeout, interval)
}

// fetchMessages returns the raw message list; the response is either a JSON
// array or an object wrapping it under "messages".
func (t *TempMailIo) fetchMessages() ([]tempMailIoItem, error) {
	if t.email == "" {
		return nil, &TempMailError{Provider: "temp-mail.io", Message: "no email — call GenerateEmail() first"}
	}
	body, err := t.authReq(http.MethodGet, "/email/"+t.email+"/messages", nil)
	if err != nil {
		return nil, &TempMailError{Provider: "temp-mail.io", Message: err.Error()}
	}
	var items []tempMailIoItem
	if e := json.Unmarshal(body, &items); e != nil {
		var wrap struct {
			Messages []tempMailIoItem `json:"messages"`
		}
		if e2 := json.Unmarshal(body, &wrap); e2 == nil {
			return wrap.Messages, nil
		}
		return nil, &TempMailError{Provider: "temp-mail.io", Message: "parse error: " + e.Error()}
	}
	return items, nil
}

type tempMailIoItem struct {
	ID          json.RawMessage `json:"id"`
	UID         json.RawMessage `json:"uid"`
	From        json.RawMessage `json:"from"`
	Subject     string          `json:"subject"`
	CreatedAt   string          `json:"created_at"`
	Date        string          `json:"date"`
	BodyText    string          `json:"body_text"`
	Text        string          `json:"text"`
	BodyHTML    string          `json:"body_html"`
	HTML        string          `json:"html"`
	Attachments []struct {
		Filename    string `json:"filename"`
		ContentType string `json:"content_type"`
		Size        int    `json:"size"`
	} `json:"attachments"`
}

File: python/tempmail_wrapper/providers/temp_mail_io.py
"""temp-mail.io provider implementation (REST API, token-based internal API)."""

from __future__ import annotations

import random
import string
from datetime import datetime

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail


class TempMailIoProvider(TempMailProvider):
    """Provider for temp-mail.io — free internal REST API, Bearer token."""

    BASE_URL = "https://api.internal.temp-mail.io/api/v3"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=False)
        self._token: str | None = None
        self._email: str | None = None

    @property
    def name(self) -> str:
        return "temp-mail.io"

    def _headers(self) -> dict[str, str]:
        h = {"Content-Type": "application/json"}
        if self._token:
            h["Authorization"] = f"Bearer {self._token}"
        return h

    def generate_email(self) -> str:
        resp = self._http.post(
            f"{self.BASE_URL}/email/new",
            json={"min_name_length": 6, "max_name_length": 12},
            headers=self._headers(),
            timeout=30,
        )
        if resp.status_code != 200:
            raise TempMailError(
                f"Failed to create email: {resp.status_code} {resp.text[:200]}",
                self.name,
            )
        data = resp.json()
        self._email = data.get("email")
        self._token = data.get("token")
        if not self._email:
            raise TempMailError("Missing email in response", self.name)
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        resp = self._http.get(
            f"{self.BASE_URL}/email/{email}/messages",
            headers=self._headers(),
            timeout=30,
        )
        if resp.status_code != 200:
            raise TempMailError(f"Inbox error: {resp.status_code}", self.name)
        data = resp.json()
        if not isinstance(data, list):
            data = data.get("messages", data.get("mails", []))
        messages = []
        for item in data:
            date_str = item.get("created_at", item.get("date", ""))
            try:
                date = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
            except (ValueError, AttributeError):
                date = datetime.now()
            sender = item.get("from", "")
            if isinstance(sender, dict):
                sender = sender.get("address", sender.get("name", ""))
            messages.append(
                Message(
                    id=str(item.get("id", item.get("uid", ""))),
                    sender=sender,
                    subject=item.get("subject", ""),
                    date=date,
                )
            )
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        # temp-mail.io messages endpoint returns full message list
        # Individual message read: GET /api/v3/message/{id}
        if not self._email:
            raise TempMailError("No email — call generate_email() first", self.name)
        resp = self._http.get(
            f"{self.BASE_URL}/email/{self._email}/messages",
            headers=self._headers(),
            timeout=30,
        )
        if resp.status_code != 200:
            raise NotFoundError(f"Cannot read inbox: {resp.status_code}", self.name)
        data = resp.json()
        if not isinstance(data, list):
            data = data.get("messages", data.get("mails", []))
        for item in data:
            if str(item.get("id", item.get("uid", ""))) == str(message_id):
                date_str = item.get("created_at", item.get("date", ""))
                try:
                    date = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
                except (ValueError, AttributeError):
                    date = datetime.now()
                sender = item.get("from", "")
                if isinstance(sender, dict):
                    sender = sender.get("address", sender.get("name", ""))
                return MessageDetail(
                    id=str(message_id),
                    sender=sender,
                    subject=item.get("subject", ""),
                    date=date,
                    body_text=item.get("body_text", item.get("text", "")),
                    body_html=item.get("body_html", item.get("html", "")),
                    attachments=[
                        {
                            "filename": a.get("filename", ""),
                            "content_type": a.get("content_type", ""),
                            "size": a.get("size", 0),
                        }
                        for a in item.get("attachments", [])
                    ],
                )
        raise NotFoundError(f"Message {message_id} not found", self.name)

    def delete_email(self, email: str) -> bool:
        self._token = None
        self._email = None
        return True
